pkg/lib/cockpit.d.ts

```diff
@@ -212,12 +214,12 @@ declare module 'cockpit' {
     function ngettext(context: string, message1: string, messageN: string, n: number): string;
 
     function format(format_string: string, ...args: unknown[]): string;
-    function format_number(n: number, precision?: number): string
-    function format_bytes(n: number, factor?: 1000 | 1024, options?: FormatOptions & { separate?: false }): string;
-    function format_bytes(n: number, factor: 1000 | 1024, options: FormatOptions & { separate: true }): string[];
-    function format_bytes_per_sec(n: number, factor?: 1000 | 1024, options?: FormatOptions & { separate?: false }): string;
-    function format_bytes_per_sec(n: number, factor: 1000 | 1024, options: FormatOptions & { separate: true }): string[];
-    function format_bits_per_sec(n: number, factor?: 1000 | 1024, options?: FormatOptions & { separate?: false }): string;
-    function format_bits_per_sec(n: number, factor: 1000 | 1024, options: FormatOptions & { separate: true }): string[];
-    function get_byte_units(guide_value: number, factor?: 1000 | 1024): ByteUnit[];
+    function format_number(n: number | null | undefined, precision?: number): string
+    function format_bytes(n: number | null | undefined, factor?: 1000 | 1024 | string, options?: FormatOptions & { separate?: false } | false): string;
+    function format_bytes(n: number | null | undefined, factor: 1000 | 1024 | string | undefined, options: FormatOptions & { separate: true } | true): string[];
```

Comment:
I want to veto this, this is a terrible API. We should not type how we don't want code to be written.

Passing factor as `MB` and then reverse looking up that it was 1000 is awful imo and we should rid ourselves of it.

Reply:
I can accept this argument, but it leaves us in an interesting place: we test this behaviour, but want to make it statically impossible...

Reply:
One way we can do this is with `as any` casts in the tests... not sure how I feel about it...

Reply:
My believe is that typing should help us uncover weird things, like the `superuser: true` and we should then fix this

Reply:
Let's chat about this tomorrow.  There's definitely a conversation to be had here...

Reply:
`n: number | null | undefined` is legit. We [document it](https://cockpit-project.org/guide/latest/cockpit-util.html), and it's a sensible simplification/robustification/convenience of JS code for showing e.g. optional D-Bus/VM/container/etc. properties.

But AFAIUI @jelly's concern is not about that, but about `factor`, right? The docs say:

> If specifying 1000 or 1024 is specified as a factor then an appropriate suffix will be chosen. By default the factor is 1000. You can pass a string suffix as a factor in which case the resulting number will be formatted with the same suffix.

Not too long ago I made some effort to move almost all numbers in Cockpit to SI base-10 units, like network speeds or disk sizes (as reality/hardware is base-10), with RAM size being the *only* legit exception:
```
pkg/systemd/hw-detect.js:            memorySize = cockpit.format_bytes(memorySize, 1024);
```

The intention was to move as many format_bytes*() calls as possible to a single argument and the default representation. There are a few stragglers like
```
pkg/metrics/metrics.jsx:                                        label={ cockpit.format(_("$0 free"), cockpit.format_bytes(info.avail, 1000)) } />
pkg/metrics/metrics.jsx:                                        content={ cockpit.format(_("$0 total"), cockpit.format_bytes(info.size, 1000)) }
```
where we can just drop the `1000`.

The Storage page shows both:
```
pkg/storaged/utils.js:    const with_decimal_unit = cockpit.format_bytes(bytes, 1000);
pkg/storaged/utils.js:    const with_binary_unit = cockpit.format_bytes(bytes, 1024);
```

I checked our related projects (everything on https://cockpit-project.org/applications.html). Outside of cockpit.git, there is just a single invocation of  `format_bytes_per_sec()` (in the old 45drives -navigator) with a single argument :white_check_mark: .

Most invocations of format_bytes() are just single-argument (the number), with only two exceptions:

 - https://github.com/cockpit-project/cockpit-podman/pull/1672 (harmless)
 - podman's [format_memory_and_limit()](https://github.com/cockpit-project/cockpit-podman/blob/8f538002ba8831d401f44a8b638718c3cfbc602a/src/util.js#L47) actively uses the "string" functionality. The intention is good, and it's actually really difficult to achieve exactly this in another way, but perhaps we have an idea. It's only used in a single place, and I wouldn't mind showing the container memory usage as a percentage, like "10% of 500 MB" instead of "50 / 500 MB".
 - That format_memory_and_limit() function also uses a deprecated and undocumented boolean third argument, which used to be `separate`. This has become an option now. I fixed this in the same PR above. 

So: We are actually at liberty to change this API, as (1) the only non-trivial consumer is under our control, and (2) cockpit-podman even bundles cockpit.js, so it wouldn't even be a runtime API break.

If I were to do it from scratch, I would completely drop `factor`, and add a `binary` or perhaps `base2` bool flag to FormatOptions, to clean up the API and point out more strongly that we strongly prefer base-10. We could even add a backwards compatibility hack to check if `options` is not a dict, the next argument is, and then ignore that argument  (as it is the deprecated "format" one). That'll cause some interesting annotations in the unit test, but that's worth it IMHO.

AFAICS, that "options can be a bool" old API isn't even covered here? We do cover it in the unit tests:
```
    // backwards compatible API: format_bytes with a boolean options (used to be a single "separate" flag)
    assert.strictEqual(cockpit.format_bytes(2500000, 1000, false), "2.50 MB");
    assert.deepEqual(cockpit.format_bytes(2500000, 1000, true), ["2.50", "MB"]);
```
but I'd be more than happy to drop that now (should be safe, see above). Also, in fact I wouldn't mind dropping the entire "separate" option -- as shown above, it's only being used in a single place, and I'd really like to get rid of it -- it encourages playing dirty tricks with string mangling which may be hard or impossible to translate.

So if we can get rid of that single podman usage, we can end up with
```
string = cockpit.format_bytes(n: number | null | undefined, options?: FormatOptions): string
```

where the available options would be `precision: int` and `binary: bool` (or `base2`). That seems quite sane to me then?

@allisonkarlitskaya @jelly WDYT?

Reply:
I sent https://github.com/cockpit-project/cockpit-podman/pull/1673 to drop the last non-trivial usage from podman. If/when that goes in, I'm happy to fix the only other consumer of that (overview memory usage)

Reply:
Just to be clear, "get rid of" could mean:
 - completely remove this functionality from `cockpit.js`, breaking the API
 - remove the advertisement of this functionality from `cockpit.d.ts`, breaking only typescript users
 - `/** @deprecate */` the functionality in `cockpit.d.ts`
 - something else?

Reply:
I'd *like* to actually break the API, given that there are zero *known* users; but I'm too scared about the unknown ones. So, my gut feeling:

 - Design the API as we *want* it (see my proposal above, I'm happy to discuss of course)
 - Do the types exactly for this, so that we learn about unintended API usage via type checks
 - Then add some backwards compat hacks inside of `format_bytes()` to opportunistically interpret a non-object second argument: if it's "1024", treat it as `{ binary: true }`, and ignore all others. console.warn() about it in any case.

Is that possible/reasonable?

Reply:
So here's one proposal that I think you might like, which would let us keep something like the existing podman case:

```typescript
    /* === Number formatting ===================== */

    interface FormatFactor { _tag: "format-factor" }
    type FormatOptions = {
        output?: "value" | "unit";
        precision?: number;
        base2?: boolean;
        factor?: FormatFactor;
    };

    function format_number(n?: number?, precision?: number): string
    function choose_format_factor(n?: number): FormatFactor;
    function format_bytes(n?: number?, options?: FormatOptions): string;
    function format_bytes_per_sec(n?: number?, options?: FormatOptions): string;
    function format_bits_per_sec(n?: number?, options?: FormatOptions): string;
```

and here's a simplified version that makes the podman case impossible:

```typescript
    /* === Number formatting ===================== */

    type FormatOptions = {
        precision?: number;
        base2?: boolean;
    };

    function format_number(n?: number?, precision?: number): string
    function format_bytes(n?: number?, options?: FormatOptions): string;
    function format_bytes_per_sec(n?: number?, options?: FormatOptions): string;
    function format_bits_per_sec(n?: number?, options?: FormatOptions): string;
```

I'd be OK with either.

Reply:
TBH I don't understand how the first proposal would work for "the podman case", by which I suppose you mean "uses the `separate` option" -- it will return a `string[]` then (or even before https://github.com/cockpit-project/cockpit-podman/pull/1672 it would get a boolean third argument). 

Also, for the second one, `options` is *really* an object (`{ precision: 3 }`), whereas in the first one the `{}` feel more like the grouping for the various types that the second argument could have? (bool or int or string). This doesn't seem consistent?

(That said, I really dislike the current API; `bool | int | string` is just crazy)

Reply:
The idea with the more complicated API is that you can use `choose_format_factor(size)` to get an opaque `FormatFactor` object which can then be passed as `factor` to multiple invocations of `format_bytes()` to force the same units from each invocation.  You could then opt out of the display of the units with `output: "value";`

It was just an idea.  I think I'd actually prefer not to do it.